### PR TITLE
ci: rewrite test build pipeline in python

### PR DESCRIPTION
**Pull request by benesch**
This makes things simpler, since the logic is not split between Bash and
Python, and more robust, because Python has actual error handling.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/2696)

### CODE REVIEW

File: misc/python/materialize/bintray.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+"""
+
+from typing import IO, Sequence
+from typing_extensions import Final
+from requests import Session, Response
+import requests
+
+API_BASE: Final = "https://api.bintray.com"
+
+HTTPError = requests.HTTPError
```
Comment (quodlibetor) [inline misc/python/materialize/bintray.py:25]: Is there a reason this isn't part of `from requests import ..`?

File: misc/python/materialize/bintray.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+
+class VersionAlreadyExistsError(Exception):
+    """Raised when creating a version of a Bintray package that already
+    exists."""
+
+    pass
```
Comment (quodlibetor) [inline misc/python/materialize/bintray.py:32]: technically the docstring is the first statement in this class and the `pass` isn't necessary, but :shrug: it just saves some vertical space and I think people occasionally find it confusing if the pass is missing.

File: misc/python/materialize/bintray.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+
+class PackageClient:
+    """A client for a specific Bintray package."""
+
+    def __init__(self, session: Session, subject: str, repo: str, package: str):
```
Comment (quodlibetor) [inline misc/python/materialize/bintray.py:38]: this should `-> None` to get mypy to typecheck it. 😢

Comment (quodlibetor) [reply misc/python/materialize/bintray.py:38]: I think that if `deny_untyped_defs` is turned on in the mypy configuration then you'll get errors for things that aren't checked.

Comment (benesch) [reply misc/python/materialize/bintray.py:38]: `deny_untyped_defs` explodes horribly on decoding JSON and YAML unfortunately. Haven't figured out a good way around that yet. Do you have suggestions? There doesn't seem to be much prior art on typed JSON/YAML decoding in Python.

File: misc/python/materialize/bintray.py (hunk 1 above)
Comment (benesch) [reply misc/python/materialize/bintray.py:25]: Oops, dead code. Thanks. Really miss Rust's lint on that.

File: misc/python/materialize/bintray.py (hunk 2 above)
Comment (benesch) [reply misc/python/materialize/bintray.py:32]: Ah, thanks! Docstring was added later. Fixed.

**Comment by benesch** [comment]
How weird. The build didn't get triggered. Maybe GitHub flakiness?

**Pull request merged**